src/hotspot/share/gc/shared/slidingForwarding.inline.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2021, Red Hat, Inc. All rights reserved.
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_GC_SHARED_SLIDINGFORWARDING_INLINE_HPP
+#define SHARE_GC_SHARED_SLIDINGFORWARDING_INLINE_HPP
+
+#ifdef _LP64
+
+#include "gc/shared/slidingForwarding.hpp"
+#include "oops/markWord.hpp"
+#include "oops/oop.inline.hpp"
+#include "utilities/macros.hpp"
+
+size_t SlidingForwarding::region_index_containing(HeapWord* addr) const {
+  assert(addr >= _heap_start, "sanity: addr: " PTR_FORMAT " heap base: " PTR_FORMAT, p2i(addr), p2i(_heap_start));
+  size_t index = ((size_t) (addr - _heap_start)) >> _region_size_words_shift;
+  assert(index < _num_regions, "Region index is in bounds: " PTR_FORMAT, p2i(addr));
+  return index;
+}
+
+bool SlidingForwarding::region_contains(HeapWord* region_base, HeapWord* addr) const {
+  return uintptr_t(addr - region_base) < (1ull << _region_size_words_shift);
+}
+
+
+uintptr_t SlidingForwarding::encode_forwarding(HeapWord* original, HeapWord* target) {
+  size_t orig_idx = region_index_containing(original);
+  size_t base_table_idx = orig_idx * 2;
+  size_t target_idx = region_index_containing(target);
+  HeapWord* encode_base;
+  uintptr_t region_idx;
+  for (region_idx = 0; region_idx < NUM_TARGET_REGIONS; region_idx++) {
+    encode_base = _target_base_table[base_table_idx + region_idx];
+    if (encode_base == UNUSED_BASE) {
+      encode_base = _heap_start + target_idx * (1ull << _region_size_words_shift);
+      _target_base_table[base_table_idx + region_idx] = encode_base;
+      break;
+    } else if (region_contains(encode_base, target)) {
+      break;
+    }
+  }
+  if (region_idx >= NUM_TARGET_REGIONS) {
+    assert(G1GC_ONLY(UseG1GC) NOT_G1GC(false), "Only happens with G1 serial compaction");
```

Comment:
I think the "top" GC flags are always available -- they are deliberately defined in shared `gc_globals.hpp`, so you can just `assert(UseG1GC, ...` here?